js/private/js_image_layer.bzl

```diff
@@ -0,0 +1,226 @@
+
+def _runfiles_dir(root, default_info):
+    manifest = default_info.files_to_run.runfiles_manifest
+
+    runfiles = manifest.short_path.replace(manifest.basename, "")[:-1]
+    return paths.join(root, runfiles.replace(".sh", ""))
+
+def _js_image_layer_impl(ctx):
+    if len(ctx.attr.binary) != 1:
+        fail("binary attribute has more than one transition")
+
+    default_info = ctx.attr.binary[0][DefaultInfo]
+    runfiles_dir = _runfiles_dir(ctx.attr.root, default_info)
+
+    executable = default_info.files_to_run.executable
+    executable_path = paths.replace_extension(paths.join(ctx.attr.root, executable.short_path), "")
+    real_executable_path = _runfile_path(ctx, executable, runfiles_dir)
+    launcher = _write_laucher(ctx, real_executable_path)
+
+    files = {}
+
+    files[executable_path] = {"dest": launcher.path, "root": launcher.root.path}
+
+    for file in depset(transitive = [default_info.files, default_info.default_runfiles.files]).to_list():
+        destination = _runfile_path(ctx, file, runfiles_dir)
+        entry = {"dest": file.path, "root": file.root.path, "is_source": file.is_source, "is_directory": file.is_directory}
+        if destination == real_executable_path:
+            entry["remove_non_hermetic_lines"] = True
+        files[destination] = entry
+
+    entries = ctx.actions.declare_file("{}_entries.json".format(ctx.label.name))
+    ctx.actions.write(entries, content = json.encode(files))
+
+    extension = ".tar"
+
+    if ctx.attr.compression == "gzip":
+        extension = ".tar.gz"
+
+    app = ctx.actions.declare_file("{name}_app{extension}".format(name = ctx.label.name, extension = extension))
+    node_modules = ctx.actions.declare_file("{name}_node_modules{extension}".format(name = ctx.label.name, extension = extension))
+
+    args = ctx.actions.args()
+    args.add(entries)
+    args.add(app)
+    args.add(node_modules)
+    args.add(ctx.attr.compression if ctx.attr.compression else "none")
+
+    ctx.actions.run(
+        inputs = depset([executable, launcher, entries], transitive = [default_info.files, default_info.default_runfiles.files]),
+        outputs = [app, node_modules],
+        arguments = [args],
+        executable = ctx.executable._builder,
+        progress_message = "JsImageLayer %{label}",
+        env = {
+            "BAZEL_BINDIR": ".",
+        },
+    )
+
+    return [
+        DefaultInfo(files = depset([app, node_modules])),
+        OutputGroupInfo(app = depset([app]), node_modules = depset([node_modules])),
+    ]
+
+def _js_image_layer_transition_impl(settings, attr):
+    # buildifier: disable=unused-variable
+    _ignore = (settings)
+    if not attr.platform:
+        return {}
+    return {
+        "//command_line_option:platforms": str(attr.platform),
+    }
+
+_js_image_layer_transition = transition(
+    implementation = _js_image_layer_transition_impl,
+    inputs = [],
+    outputs = ["//command_line_option:platforms"],
+)
+
+js_image_layer = rule(
```

Comment:
nit: to guarantee users can write custom rules that re-use our logic, you should also export a struct so that the attrs list and implementation function is accessible. See all the examples in bazel-lib